### PR TITLE
[4.1] RavenDB-11031 

### PR DESCRIPTION
**Pull request by aviv86**

### CODE REVIEW

File: src/Raven.Server/Documents/Handlers/CountersHandler.cs [hunk 1]
```diff
@@ -109,6 +107,7 @@ public override int Execute(DocumentsOperationContext context)
                                 LoadDocument();
                                  _database.DocumentsStorage.CountersStorage.PutCounterFromReplication(context, kvp.Key,
                                     operation.CounterName, operation.ChangeVector, operation.Delta);
+                                LastChangeVector = operation.ChangeVector;
```
Comment (ayende) [inline src/Raven.Server/Documents/Handlers/CountersHandler.cs]: This is wrong and may reset the change vector back, actually

File: src/Raven.Server/Documents/Revisions/RevisionsStorage.cs [hunk 2]
```diff
@@ -297,6 +302,27 @@ public RevisionsCollectionConfiguration GetRevisionsConfiguration(string collect
                     }
                 }
 
+                if (configuration.IncludeCountersSnapshot &&
+                    document.TryGet(Constants.Documents.Metadata.Key, out BlittableJsonReaderObject metadata) &&
+                    metadata.TryGet(Constants.Documents.Metadata.Counters, out BlittableJsonReaderArray counterNames))
+                {
+                    var countersSnapshot = counterNames.ToDictionary(c => c.ToString(),
```
Comment (ayende) [inline src/Raven.Server/Documents/Revisions/RevisionsStorage.cs]: Let's avoid having complex nested Linq ops.
Build the `DynamicJsonObject` directly

File: src/Raven.Client/Documents/Operations/Revisions/RevisionsCollectionConfiguration.cs [hunk 3]
```diff
@@ -13,12 +13,15 @@ public class RevisionsCollectionConfiguration
 
         public bool PurgeOnDelete { get; set; }
 
+        public bool IncludeCountersSnapshot { get; set; }
```
Comment (ayende) [inline src/Raven.Client/Documents/Operations/Revisions/RevisionsCollectionConfiguration.cs]: Under what scenario I _don't_ want that?
We don't give a choice on attachments, and we shouldn't here.

File: src/Raven.Server/Documents/Handlers/CountersHandler.cs (hunk 1 above)
Comment (ayende) [reply src/Raven.Server/Documents/Handlers/CountersHandler.cs]: We need to _only_ allow it if this is an update, not if this is already merged / conflict with the existing one

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Documents/Revisions/RevisionsStorage.cs (hunk 2 above)
Comment (aviv86) [reply src/Raven.Server/Documents/Revisions/RevisionsStorage.cs]: done

File: src/Raven.Client/Documents/Operations/Revisions/RevisionsCollectionConfiguration.cs (hunk 3 above)
Comment (aviv86) [reply src/Raven.Client/Documents/Operations/Revisions/RevisionsCollectionConfiguration.cs]: removed

File: src/Raven.Server/Documents/Handlers/CountersHandler.cs (hunk 1 above)
Comment (aviv86) [reply src/Raven.Server/Documents/Handlers/CountersHandler.cs]: done

File: src/Raven.Server/Documents/Handlers/CountersHandler.cs [hunk 4]
```diff
@@ -105,9 +104,11 @@ public override int Execute(DocumentsOperationContext context)
                                 break;
                             case CounterOperationType.Put:
                                 LoadDocument();
-                                 _database.DocumentsStorage.CountersStorage.PutCounterFromReplication(context, kvp.Key,
+                                var cv = _database.DocumentsStorage.CountersStorage.PutCounterFromReplication(context, kvp.Key,
                                     operation.CounterName, operation.ChangeVector, operation.Delta);
-                                LastChangeVector = operation.ChangeVector;
+                                // set LastChangeVector only if this is an update
+                                if (cv != null)
```
Comment (ayende) [inline src/Raven.Server/Documents/Handlers/CountersHandler.cs]: Somewhat of a problem still, what is this is conflicting?

File: src/Raven.Server/Documents/Revisions/RevisionsStorage.cs [hunk 5]
```diff
@@ -302,16 +300,22 @@ public RevisionsCollectionConfiguration GetRevisionsConfiguration(string collect
...
                     metadata.TryGet(Constants.Documents.Metadata.Counters, out BlittableJsonReaderArray counterNames))
                 {
-                    var countersSnapshot = counterNames.ToDictionary(c => c.ToString(), 
-                        c => _documentsStorage.CountersStorage.GetCounterValue(context, id, c.ToString()).GetValueOrDefault());
+                    var dvj = new DynamicJsonValue();
+                    for (var i = 0; i < counterNames.Length; i++)
+                    {
+                        var counter = counterNames[i].ToString();
+                        var val = _documentsStorage.CountersStorage.GetCounterValue(context, id, counter);
+                        if (val == null)
```
Comment (ayende) [inline src/Raven.Server/Documents/Revisions/RevisionsStorage.cs:311]: can there be duplicates here?

**Review by ayende — CHANGES_REQUESTED** [review]

**Pull request merged**